### PR TITLE
Call aten ops through c10 dispatcher

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24253 [test] Benchmarks
* #23695 [wip] Fixed size arrays
* #26130 Remove unboxedAutogradKernel from c10
* **#23668 Call aten ops through c10 dispatcher**

- The eager mode frontend now calls operators who are defined in native_functions.yaml with use_c10_dispatcher: True through the c10 dispatcher and not anymore through globalATenDispatch().
- These operators aren't registered with globalAtenDispatch anymore, only on c10 now.
- Backend extensions calling globalATenDispatch().registerOp() to add their own kernels still work, this function will forward the registration to the c10 dispatcher for them.

Benchmarks: https://docs.google.com/document/d/1gpzKZcFf1JJameY1vKxF7Cloul9s6D8HKIK2_Pp1hFo/edit?usp=sharing

Differential Revision: [D16603133](https://our.internmc.facebook.com/intern/diff/D16603133/)

### CODE REVIEW

**Comment by apaszke** [comment]
Do we have any benchmarks for this change?

**Comment by smessmer** [comment]
@apaszke not yet, this is still wip and doesn't even build yet. We'll have benchmarks before I land it.

File: aten/src/ATen/function_wrapper.py
Comment (li-roy) [inline aten/src/ATen/function_wrapper.py]: Is it possible to move the variable logic inside the dispatch call?

Comment (smessmer) [reply aten/src/ATen/function_wrapper.py]: yes, planning to do this in a future PR. This is the easier change for now.

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: It seems pretty weird to me that I have to do another singleton lookup to actually call an OperatorHandle. Shouldn't the OperatorHandle know what dispatcher it's associated with?

**Comment by smessmer** [comment]
Benchmarks: https://docs.google.com/document/d/1gpzKZcFf1JJameY1vKxF7Cloul9s6D8HKIK2_Pp1hFo/edit?usp=sharing

File: aten/src/ATen/function_wrapper.py
Comment (smessmer) [reply aten/src/ATen/function_wrapper.py]: The reason is that the Dispatcher class is a facade API. All calls to the dispatcher go through the Dispatcher class currently, so it's all in one place. The `OperatorHandle` class is just a handle (basically an operator ID) without any functionality. We can consider changing that later and giving the `OperatorHandle` class more responsibility, but that's not how the API is currently implemented. In any case, probably not something we should change in this diff since that would be a bigger and unrelated API change to the dispatcher.

**Comment by smessmer** [comment]
not sure what's going on with CI, the failures look like infra failures, probably unrelated to this PR.

**Comment by ezyang** [comment]
Re benchmarks, it doesn't look like the hash table is breaking the bank, but what I see in the benchmarks is that we are paying a 2-8% tax for no good reason (because we don't actually need a hash table, all the IDs are contiguous.)

File: caffe2/core/tensor.h
Comment (ezyang) [inline caffe2/core/tensor.h]: What's going on here? These changes are not described in the PR body

File: aten/src/ATen/core/ATenDispatch.h
Comment (ezyang) [inline aten/src/ATen/core/ATenDispatch.h]: Shouldn't you assert that `overload_name_pos != std::string::npos`

File: aten/src/ATen/core/ATenDispatch.h
Comment (ezyang) [inline aten/src/ATen/core/ATenDispatch.h]: The fact that we have to do this is a big code smell to me. Why do we need `OperatorName` in the first place? The old ATenDispatch.h was OperatorName oblivious; the only thing it knew about was the overall schema string.

File: aten/src/ATen/core/ATenDispatch.h
Comment (ezyang) [inline aten/src/ATen/core/ATenDispatch.h]: you mean deregister an operator?

File: aten/src/ATen/core/ATenDispatch.h
Comment (ezyang) [inline aten/src/ATen/core/ATenDispatch.h]: The implementation strategy in this file is really weird to me. What I thought would have happened is that ATen native operators which had been moved to c10 would have had c10 registration code generated for them. Instead, what seems to be happening is that you're reusing the ATenDispatch::registerOp calls, and then shoving the c10 registration code inside those function calls. But then you have a problem where they get destructed immediately on exit of the function, so you shoved them in a vector... and then you had the ordering problem which is what this comment is about. But this rube goldberg machine seems totally unnecessary: just codegen c10 registrations top level and you're done! What am I missing?

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [reply aten/src/ATen/function_wrapper.py]: These will have to be collapsed into one call post Variable lives in TensorTypeSet in any case

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: Ugh. (No action necessary)

**Comment by ezyang** [comment]
Primary concern is that we're not actually generating c10 registrations, instead they have been jerry rigged on top of aten dispatch registrations. This might have been acceptable as an intermediate step, but they've caused multiple problems. You should do it properly instead.

File: aten/src/ATen/core/ATenDispatch.h
Comment (smessmer) [reply aten/src/ATen/core/ATenDispatch.h]: that would be a weaker assertion. The way it is currently asserts both, `!= npos` and that there is no off-by-one in the `find_first_of`.

File: aten/src/ATen/core/ATenDispatch.h
Comment (smessmer) [reply aten/src/ATen/core/ATenDispatch.h]: We need it for checking if a schema is already moved to the c10 dispatcher. This diff has to make ATenDispatch a bit more complicated for that because it has to keep compatibility with backend extensions, but that's ok because we're going to delete ATenDispatch soon.

File: aten/src/ATen/core/ATenDispatch.h
Comment (smessmer) [reply aten/src/ATen/core/ATenDispatch.h]: What you're thinking is right. We are codegen'ing c10 registrations, see the diff right previous to this one (https://github.com/pytorch/pytorch/pull/23667). This code here makes sure that backend extensions don't break until we want to break them, see the plan in https://github.com/pytorch/pytorch/issues/24132

**Review by smessmer — COMMENTED** [review]
As mentioned in the above comments, we are generating c10 registrations and the jerry rig is only to make sure we don't break the API for backend extensions. globalATenDispatch().registerOp() is never called from our codegen for these ops, it is only called directly from backend extensions.

File: caffe2/core/tensor.h
Comment (smessmer) [reply caffe2/core/tensor.h]: This excludes ATen/core/Tensor.h from the mobile build. It's weird we included it in the first place and it caused build errors for me because my added code depends on ATen things.

File: aten/src/ATen/core/ATenDispatch.h
Comment (ezyang) [reply aten/src/ATen/core/ATenDispatch.h]: Assert `overload_name_pos != std::string::npos` in addition to the current assert. I don't see how your code currently asserts it `!= npos`. You just blindly index into the string using `overload_name_end_pos`.

**Comment by smessmer** [comment]
Re benchmarks: They are not because of the hash table, I tested and compared it against a std::vector. I don't think further investigation is worth it because it only happens for operators that return void, which is a very rare case. Any operators returning an actual value don't see a regression.

**Comment by ezyang** [comment]
> As mentioned in the above comments, we are generating c10 registrations and the jerry rig is only to make sure we don't break the API for backend extensions. globalATenDispatch().registerOp() is never called from our codegen for these ops, it is only called directly from backend extensions.

Initially, there was quite a bit of pushback to the idea of only parting c10 registrations over partially, while the missing support was to be added. I don't think we managed to convince you back then that we should have full support in c10 before cutting over, and my idea was, well, if the partial registration could be done cleanly, I'd let it through to move things along. But this PR blows through the complexity budget I had in my head for partial registration. Did we establish that this PR is actually blocking for other workstreams? Because if it is not, I *strongly* prefer to ditch all of these workarounds and just invest in fixing c10 to support everything first. (By the way, it would have been a lot easier to have this conversation if we had known about the necessary hacks earlier in the process, rather than a month later when the PR is ready to go.)

**Comment by smessmer** [comment]
We need some ops earlier than others, that's why this is necessary. About the complexity, the only reason for the complexity here is that I'm actually doing what you're describing for the backend extension API - I'm keeping it working in its current state until all ops are moved to c10 and only then pull the switch for them. There would be no complexity here if we say we're also moving backend extensions op-by-op, but I think that would be the wrong call.

**Comment by gchanan** [comment]
disclaimer: I agree with @ezyang here; I don't believe this is actually a moving target, we haven't added complexity to aten ops / native functions for months AFAIR.

@smessmer: I don't understand your comment, it feels like there are details missing?

For example:
> We need some ops earlier than others, that's why this is necessary

Examples?

> All dispatcher changes we're planning currently will be done without c10 in mind and c10 will fall out of sync, making it harder and harder to actually merge the dispatchers. If we go down this route, there's a real risk that we're never going to merge and will always have 4 separate dispatchers.

Is it infeasible to develop dispatcher changes with c10 in mind?  AFAIK @ezyang is the only one making dispatcher changes -- there's not a lot of communication complexity there.  And don't you just have the same problem with the ops you haven't moved/supported yet?  What guarantees that we make dispatcher changes with those in mind?

**Comment by smessmer** [comment]
> > We need some ops earlier than others, that's why this is necessary
> 
> Examples?
> 

This is what we talked about in the last meeting. Mobile only needs a small set of ops but it needs them on the c10 dispatcher and early.

> > All dispatcher changes we're planning currently will be done without c10 in mind and c10 will fall out of sync, making it harder and harder to actually merge the dispatchers. If we go down this route, there's a real risk that we're never going to merge and will always have 4 separate dispatchers.
> 
> Is it infeasible to develop dispatcher changes with c10 in mind? AFAIK @ezyang is the only one making dispatcher changes -- there's not a lot of communication complexity there. And don't you just have the same problem with the ops you haven't moved/supported yet? What guarantees that we make dispatcher changes with those in mind?

Afaik, all of Edwards changes currently are not implementing anything for the c10 dispatcher. @ezyang correct me if I'm wrong but I think your plan was to land multi dispatch without it working in c10? I strongly think this is the wrong approach. Landing this before that will make sure all our changes work in c10 as well. This means a bit of additional overhead in implementing multi dispatch, because we have to implement it for c10 and for ATen dispatch, but that's something I've been continuously calling out even before we started working on ATen dispatch and it is the price we have to pay if we want to do multi dispatch now and still merge the dispatchers. As said before, if we implement multi dispatch without implementing it for c10, it gets unlikely that we'll be able to do the dispatcher merge.

**Comment by smessmer** [comment]
If we land this PR, it will add the ability for you to test it in c10, it will actually force you to have it correctly working in c10 because if it doesn't work, CI will fail. Right now, the c10 dispatcher and globalATenDispatch are at feature parity. If we land multi dispatch for globalATenDispatch only, that will no longer be the case.

**Comment by smessmer** [comment]
Since all the complexity comes from the backend extension API, there's one alternative. We could move the backend extension API to a separate file and keep ATenDispatch.h simple. That would, however, require backend extensions to modify their code two times instead of one time - one time now when we move the API to a new intermediate one and a second time later when we are finished moving everything to c10.

**Comment by smessmer** [comment]
@dzhulgakov mentioned another possible alternative. I could move all complexity into the c10 dispatcher. The reasons why I can't move some ops in this PR yet is that their schemas are unparseable. By making schema optional in the c10 dispatcher, I could move all ops immediately.

We'd still need the use_c10_dispatcher flag and treat things differently inside the registration calls to the c10 dispatcher, and in total that will be more complexity than it is in this current PR because the c10 dispatcher will have to deal with some ops being boxed, some unboxed with schema, some unboxed without schema, ..., but the complexity would be in a different place and ATenDispatch would be deleted immediately.

It would be a more complex implementation, but the upside would be that it's a less complex dispatcher API - registration API calls would look the same for all ops (or almost the same - some would supply a schema and some wouldn't) and only do different things internally.

I personally think it's a worse approach because it is more total complexity, but I'm fine going that route if it unblocks me from landing this. Btw, in both approaches, we should not land the multi dispatch PR while it doesn't implement it for c10. In the original approach of this PR, that PR would have to add c10 functionality, in this new proposed approach the multi dispatch PR would have to wait for this to be implemented and then only do the c10 thing.

**Comment by ezyang** [comment]
I like @dzhulgakov's proposal, however, I cannot easily assess claims about it being more complicated for c10. In particular:

> some ops being boxed, some unboxed with schema, some unboxed without schema

Can you say a little more about what the substantive difference between "unboxed with schema" and "unboxed without schema" is? In ATenDispatch.h, no schema parsing is necessary, so I don't know what concretely this changes in c10.

**Comment by smessmer** [comment]
I'd still need to do the op-by-op approach and the `use_c10_dispatcher` flag but instead of moving them op-by-op from globalATenDispatch to the c10 dispatcher, they'd all already be on the c10 dispatcher and they would be moved op-by-op from being a schemaless unboxed c10 op to being a schemaful unboxed c10 op and then to being a schemaful boxed c10 op, which is the state we need them in for mobile. So it's more cases of different types of operators that the c10 dispatcher has to handle internally.

File: aten/src/ATen/core/TensorMethods.h
Comment (dzhulgakov) [inline aten/src/ATen/core/TensorMethods.h]: why is variable switch lives in each and every method? can't it live inside dispatcher? it'd make Ed's changes to multidispatch easier

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/core/TensorMethods.h
Comment (smessmer) [reply aten/src/ATen/core/TensorMethods.h]: this is now gone and uses VariableTensorId

**Pull request closed without merging**
